File: projects/iffywow/elixir/lib/ithkuil/coord.ex
defmodule Ithkuil.Coord do
  @moduledoc """
  The canonical, versioned coordinate tuple — the single source of truth for a
  New Ithkuil word in this SDK (see `../../README.md` and `../../CODEC.md`).

  Canonical Elixir-internal form:

      {:ithkuil_word, version,
       [
         {:glyph, character_class, base, orientation,
          [
            {socket_id, {:modifier, shape, orientation, diacritics, child_sockets}}
          ]}
       ]}

  Canonical-form invariants (CODEC.md §2):

    * `version >= 1`
    * `orientation in 0..3`, `socket_id in 0..7` (schema v1)
    * sockets sorted strictly ascending by socket id; duplicates rejected
    * **empty sockets are omitted** — `{id, :empty}` / wire `[id, null]` are
      accepted on input but never appear in the canonical form

  Wire form (JSON tagged arrays, as plain Elixir lists — JSON text is the
  caller's concern):

      ["ithkuil-word", v, glyphs]
      ["glyph", character_class, base, orientation, sockets]
      [socket_id, nil | modifier]
      ["modifier", shape, orientation, diacritics, sockets]

  Errors are `{:error, class}` or `{:error, {class, detail}}` where `class`
  matches the conformance vocabulary (`:invalid_version`,
  `:invalid_orientation`, `:invalid_socket_id`, `:duplicate_socket`, ...).
  """

  @word_tag "ithkuil-word"
  @glyph_tag "glyph"
  @modifier_tag "modifier"
  @max_socket_id 7

  @type socket_id :: 0..7
  @type orientation :: 0..3
  @type diacritic :: non_neg_integer()
  @type modifier ::
          {:modifier, shape :: non_neg_integer(), orientation(), [diacritic()], [socket()]}
  @type socket :: {socket_id(), modifier()}
  @type loose_socket :: {non_neg_integer(), modifier() | :empty | nil}
  @type glyph ::
          {:glyph, character_class :: non_neg_integer(), base :: non_neg_integer(),
           orientation(), [socket()]}
  @type t :: {:ithkuil_word, pos_integer(), [glyph()]}
  @type wire :: [term()]
  @type reason :: atom() | {atom(), term()}

  @doc "Wire tag strings (positional in the byte encoding; strings exist only in JSON)."
  def word_tag, do: @word_tag
  def glyph_tag, do: @glyph_tag
  def modifier_tag, do: @modifier_tag

  @doc """
  Build a canonical coordinate from either the wire form (tagged arrays) or a
  (possibly loose) internal tuple. Returns `{:ok, t}` or `{:error, reason}`.
  """
  @spec new(wire() | tuple()) :: {:ok, t()} | {:error, reason()}
  def new(input) when is_list(input), do: from_wire(input)
  def new({:ithkuil_word, _, _} = coord), do: canonicalize(coord)
  def new(other), do: {:error, {:invalid_structure, other}}

  @doc """
  Check that `coord` is already in canonical form.
  Returns `:ok` or `{:error, reason}` (`:not_canonical` when merely non-canonical).
  """
  @spec validate(term()) :: :ok | {:error, reason()}
  def validate(coord) do
    case canonicalize(coord) do
      {:ok, ^coord} -> :ok
      {:ok, _other} -> {:error, :not_canonical}
      {:error, reason} -> {:error, reason}
    end
  end

  @doc """
  Validate and canonicalize a (loose) internal tuple: sockets are sorted
  ascending, duplicate socket ids rejected, `{id, :empty}` / `{id, nil}`
  entries dropped.
  """
  @spec canonicalize(term()) :: {:ok, t()} | {:error, reason()}
  def canonicalize({:ithkuil_word, version, glyphs}) when is_list(glyphs) do
    if is_integer(version) and version >= 1 do
      with {:ok, gs} <- map_all(glyphs, &canon_glyph/1) do
        {:ok, {:ithkuil_word, version, gs}}
      end
    else
      {:error, {:invalid_version, version}}
    end
  end

  def canonicalize(other), do: {:error, {:invalid_structure, other}}

  @doc """
  Parse the JSON-tagged-array wire form (already decoded to Elixir lists /
  strings / integers / nil) into a canonical coordinate.
  """
  @spec from_wire(term()) :: {:ok, t()} | {:error, reason()}
  def from_wire(wire) do
    with {:ok, loose} <- wire_to_tuple(wire), do: canonicalize(loose)
  end

  @doc "Canonical coordinate -> wire form (tagged arrays as Elixir data)."
  @spec to_wire(t()) :: wire()
  def to_wire({:ithkuil_word, version, glyphs}) do
    [@word_tag, version, Enum.map(glyphs, &glyph_to_wire/1)]
  end

  # -- wire -> loose tuple (structure only; value rules live in canonicalize) --

  defp wire_to_tuple([@word_tag, v, glyphs]) when is_list(glyphs) do
    with {:ok, gs} <- map_all(glyphs, &glyph_from_wire/1), do: {:ok, {:ithkuil_word, v, gs}}
  end

  defp wire_to_tuple(other), do: {:error, {:invalid_structure, other}}

  defp glyph_from_wire([@glyph_tag, cc, base, o, sockets]) when is_list(sockets) do
    with {:ok, socks} <- map_all(sockets, &socket_from_wire/1) do
      {:ok, {:glyph, cc, base, o, socks}}
    end
  end

  defp glyph_from_wire(other), do: {:error, {:invalid_structure, other}}

  defp socket_from_wire([id, nil]), do: {:ok, {id, :empty}}

  defp socket_from_wire([id, modifier]) do
    with {:ok, m} <- modifier_from_wire(modifier), do: {:ok, {id, m}}
  end

  defp socket_from_wire(other), do: {:error, {:invalid_structure, other}}

  defp modifier_from_wire([@modifier_tag, shape, o, diacritics, sockets])
       when is_list(diacritics) and is_list(sockets) do
    with {:ok, socks} <- map_all(sockets, &socket_from_wire/1) do
      {:ok, {:modifier, shape, o, diacritics, socks}}
    end
  end

  defp modifier_from_wire(other), do: {:error, {:invalid_structure, other}}

  # -- canonical tuple -> wire --

  defp glyph_to_wire({:glyph, cc, base, o, sockets}) do
    [@glyph_tag, cc, base, o, Enum.map(sockets, &socket_to_wire/1)]
  end

  defp socket_to_wire({id, modifier}), do: [id, modifier_to_wire(modifier)]

  defp modifier_to_wire({:modifier, shape, o, diacritics, sockets}) do
    [@modifier_tag, shape, o, diacritics, Enum.map(sockets, &socket_to_wire/1)]
  end

  # -- canonicalization internals --

  defp canon_glyph({:glyph, cc, base, orientation, sockets}) do
    with :ok <- check_nat(cc),
         :ok <- check_nat(base),
         :ok <- check_orientation(orientation),
         {:ok, socks} <- canon_sockets(sockets) do
      {:ok, {:glyph, cc, base, orientation, socks}}
    end
  end

  defp canon_glyph(other), do: {:error, {:invalid_structure, other}}

  defp canon_sockets(sockets) when is_list(sockets), do: canon_sockets(sockets, MapSet.new(), [])
  defp canon_sockets(other), do: {:error, {:invalid_structure, other}}

  defp canon_sockets([], _seen, acc) do
    canonical =
      acc
      |> Enum.reverse()
      |> Enum.reject(fn {_id, m} -> m == :empty end)
      |> Enum.sort_by(fn {id, _m} -> id end)

    {:ok, canonical}
  end

  defp canon_sockets([{id, payload} | rest], seen, acc) do
    cond do
      not is_integer(id) or id < 0 or id > @max_socket_id ->
        {:error, {:invalid_socket_id, id}}

      MapSet.member?(seen, id) ->
        # Duplicates are invalid even when one of them is empty (one meaning,
        # one representation — mirrors web/src/coordinate.js).
        {:error, {:duplicate_socket, id}}

      payload in [:empty, nil] ->
        canon_sockets(rest, MapSet.put(seen, id), [{id, :empty} | acc])

      true ->
        with {:ok, m} <- canon_modifier(payload) do
          canon_sockets(rest, MapSet.put(seen, id), [{id, m} | acc])
        end
    end
  end

  defp canon_sockets([other | _rest], _seen, _acc), do: {:error, {:invalid_structure, other}}

  defp canon_modifier({:modifier, shape, orientation, diacritics, sockets}) do
    with :ok <- check_nat(shape),
         :ok <- check_orientation(orientation),
         :ok <- check_diacritics(diacritics),
         {:ok, socks} <- canon_sockets(sockets) do
      {:ok, {:modifier, shape, orientation, diacritics, socks}}
    end
  end

  defp canon_modifier(other), do: {:error, {:invalid_structure, other}}

  defp check_nat(n) when is_integer(n) and n >= 0, do: :ok
  defp check_nat(n), do: {:error, {:invalid_natural_number, n}}

  defp check_orientation(o) when is_integer(o) and o >= 0 and o <= 3, do: :ok
  defp check_orientation(o), do: {:error, {:invalid_orientation, o}}

  defp check_diacritics(diacritics) when is_list(diacritics) do
    if Enum.all?(diacritics, &(is_integer(&1) and &1 >= 0)) do
      :ok
    else
      {:error, {:invalid_natural_number, diacritics}}
    end
  end

  defp check_diacritics(other), do: {:error, {:invalid_structure, other}}

  defp map_all(items, fun) do
    items
    |> Enum.reduce_while({:ok, []}, fn item, {:ok, acc} ->
      case fun.(item) do
        {:ok, v} -> {:cont, {:ok, [v | acc]}}
        {:error, _} = e -> {:halt, e}
      end
    end)
    |> case do
      {:ok, acc} -> {:ok, Enum.reverse(acc)}
      error -> error
    end
  end
end

File: projects/iffywow/python/ithkuil/coord.py
"""Canonical coordinate tuple: frozen dataclasses + wire-form conversion.

Mirrors ``web/src/coordinate.js`` and CODEC.md section 2 (schema v1).

Canonical tuple (Python form)::

    Word(version, (Glyph(character_class, base, orientation, sockets), ...))
    Socket(socket_id, Modifier(shape, orientation, diacritics, sockets))

Canonical-form invariants (enforced by the dataclass constructors):

* ``version >= 1``; ``orientation in 0..3``; ``socket_id in 0..7``.
* Sockets sorted strictly ascending by ``socket_id`` (duplicates invalid).
* EMPTY SOCKETS ARE OMITTED. There is no "empty" Socket value; a Socket
  always carries a Modifier. The wire form ``[id, null]`` is accepted on
  *input* at the JSON boundary (``word_from_wire``) and dropped during
  canonicalization -- one meaning, one representation.

JSON wire form (tagged arrays, order-preserving)::

    ["ithkuil-word", version, [glyph, ...]]
    ["glyph", character_class, base, orientation, [socket, ...]]
    [socket_id, null | modifier]
    ["modifier", shape, orientation, [diacritic_id, ...], [socket, ...]]

All rejection paths raise :class:`IthkuilError` (a ``ValueError`` subclass)
whose message starts with a stable machine-readable code, e.g.
``"invalid_orientation: invalid coordinate at $[2][0][3]: orientation must
be 0..3"``. The codes are shared with ``conformance/invalid_inputs.jsonl``.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, List, Optional, Tuple, Union

WORD_TAG = "ithkuil-word"
GLYPH_TAG = "glyph"
MODIFIER_TAG = "modifier"
MAX_SOCKET_ID = 7
MAX_ORIENTATION = 3


class IthkuilError(ValueError):
    """``ValueError`` carrying a stable, machine-readable error code.

    ``str(err)`` always begins with ``"<code>: "`` so conformance tests can
    assert on either the ``code`` attribute or the message text.
    """

    def __init__(self, code: str, message: str) -> None:
        super().__init__(f"{code}: {message}")
        self.code = code


# --------------------------------------------------------------------------
# Constructor-level validation helpers (strict: canonical values only).
# --------------------------------------------------------------------------


def _require_nat(value: Any, what: str) -> int:
    if isinstance(value, bool) or not isinstance(value, int) or value < 0:
        raise IthkuilError(
            "invalid_structure", f"{what} must be a non-negative integer, got {value!r}"
        )
    return value


def _require_orientation(value: Any, what: str) -> int:
    n = _require_nat(value, what)
    if n > MAX_ORIENTATION:
        raise IthkuilError("invalid_orientation", f"{what} must be 0..{MAX_ORIENTATION}, got {n}")
    return n


def _require_sockets(items: Any, what: str) -> Tuple["Socket", ...]:
    sockets = tuple(items)
    prev = -1
    for s in sockets:
        if not isinstance(s, Socket):
            raise IthkuilError("invalid_structure", f"{what} must contain Socket instances")
        if s.socket_id == prev:
            raise IthkuilError("duplicate_socket", f"duplicate socket {s.socket_id}")
        if s.socket_id < prev:
            raise IthkuilError(
                "unsorted_sockets",
                f"sockets must be sorted strictly ascending (socket {s.socket_id} after {prev})",
            )
        prev = s.socket_id
    return sockets


# --------------------------------------------------------------------------
# Frozen dataclasses -- the canonical tuple.
# Constructors accept any sequence for the collection fields (coerced to
# tuples) but are otherwise STRICT: values must already be canonical.
# Lenient parsing/canonicalization lives in `word_from_wire`.
# --------------------------------------------------------------------------


@dataclass(frozen=True)
class Modifier:
    shape: int
    orientation: int
    diacritics: Tuple[int, ...] = ()
    sockets: Tuple["Socket", ...] = ()

    def __post_init__(self) -> None:
        _require_nat(self.shape, "modifier shape")
        _require_orientation(self.orientation, "modifier orientation")
        object.__setattr__(self, "diacritics", tuple(self.diacritics))
        for d in self.diacritics:
            _require_nat(d, "diacritic id")
        object.__setattr__(self, "sockets", _require_sockets(self.sockets, "modifier sockets"))


@dataclass(frozen=True)
class Socket:
    socket_id: int
    modifier: Modifier

    def __post_init__(self) -> None:
        _require_nat(self.socket_id, "socket id")
        if self.socket_id > MAX_SOCKET_ID:
            raise IthkuilError(
                "invalid_socket_id",
                f"socket id must be 0..{MAX_SOCKET_ID} in schema v1, got {self.socket_id}",
            )
        if not isinstance(self.modifier, Modifier):
            raise IthkuilError(
                "invalid_structure",
                "socket modifier must be a Modifier "
                "(empty sockets are omitted from the canonical form)",
            )


@dataclass(frozen=True)
class Glyph:
    character_class: int
    base: int
    orientation: int
    sockets: Tuple[Socket, ...] = ()

    def __post_init__(self) -> None:
        _require_nat(self.character_class, "glyph character class")
        _require_nat(self.base, "glyph base")
        _require_orientation(self.orientation, "glyph orientation")
        object.__setattr__(self, "sockets", _require_sockets(self.sockets, "glyph sockets"))


@dataclass(frozen=True)
class Word:
    version: int
    glyphs: Tuple[Glyph, ...] = ()

    def __post_init__(self) -> None:
        if isinstance(self.version, bool) or not isinstance(self.version, int) or self.version < 1:
            raise IthkuilError(
                "invalid_version", f"schema version must be a positive integer, got {self.version!r}"
            )
        glyphs = tuple(self.glyphs)
        for g in glyphs:
            if not isinstance(g, Glyph):
                raise IthkuilError("invalid_structure", "word glyphs must be Glyph instances")
        object.__setattr__(self, "glyphs", glyphs)


# --------------------------------------------------------------------------
# Wire parsing (lenient input -> canonical Word).  Mirrors coordinate.js.
# --------------------------------------------------------------------------


def _fail(code: str, msg: str, path: str) -> None:
    raise IthkuilError(code, f"invalid coordinate at {path}: {msg}")


def _nat(value: Any, path: str) -> int:
    """Non-negative integer at a JSON boundary.

    JS ``Number.isInteger(2.0)`` is true (JSON cannot distinguish 2.0 from
    2), so integral floats are accepted and normalized to int.  Booleans are
    rejected (JS ``Number.isInteger(true)`` is false).
    """
    if isinstance(value, bool):
        _fail("invalid_structure", "expected non-negative integer", path)
    if isinstance(value, int):
        if value < 0:
            _fail("invalid_structure", "expected non-negative integer", path)
        return value
    if isinstance(value, float) and value.is_integer() and value >= 0:
        return int(value)
    _fail("invalid_structure", "expected non-negative integer", path)
    raise AssertionError("unreachable")


def _orient(value: Any, path: str) -> int:
    n = _nat(value, path)
    if n > MAX_ORIENTATION:
        _fail("invalid_orientation", f"orientation must be 0..{MAX_ORIENTATION}", path)
    return n


def word_from_wire(value: Union[str, list, "Word"]) -> Word:
    """Parse a wire coordinate (tagged arrays or JSON text) into a canonical
    :class:`Word`: sockets sorted ascending, duplicates rejected, empty
    sockets (``[id, null]``) dropped.  A ``Word`` passes through unchanged.
    """
    if isinstance(value, Word):
        return value
    if isinstance(value, str):
        try:
            value = json.loads(value)
        except json.JSONDecodeError as e:
            raise IthkuilError("invalid_json", f"invalid coordinate: not JSON ({e})") from None
    if not isinstance(value, list) or len(value) != 3 or value[0] != WORD_TAG:
        _fail("invalid_structure", f'expected ["{WORD_TAG}", version, glyphs]', "$")
    version = value[1]
    if isinstance(version, float) and not isinstance(version, bool) and version.is_integer():
        version = int(version)
    if isinstance(version, bool) or not isinstance(version, int) or version < 1:
        _fail("invalid_version", "schema version must be a positive integer", "$[1]")
    if not isinstance(value[2], list):
        _fail("invalid_structure", "glyph list must be an array", "$[2]")
    glyphs = tuple(_parse_glyph(g, f"$[2][{i}]") for i, g in enumerate(value[2]))
    return Word(version, glyphs)


def _parse_glyph(g: Any, path: str) -> Glyph:
    if not isinstance(g, list) or len(g) != 5 or g[0] != GLYPH_TAG:
        _fail("invalid_structure", f'expected ["{GLYPH_TAG}", class, base, orientation, sockets]', path)
    return Glyph(
        _nat(g[1], f"{path}[1]"),
        _nat(g[2], f"{path}[2]"),
        _orient(g[3], f"{path}[3]"),
        _parse_sockets(g[4], f"{path}[4]"),
    )


def _parse_sockets(items: Any, path: str) -> Tuple[Socket, ...]:
    if not isinstance(items, list):
        _fail("invalid_structure", "sockets must be an array", path)
    seen = set()
    parsed: List[Tuple[int, Optional[Modifier]]] = []
    for i, s in enumerate(items):
        p = f"{path}[{i}]"
        if not isinstance(s, list) or len(s) != 2:
            _fail("invalid_structure", "expected [socket_id, null | modifier]", p)
        sid = _nat(s[0], f"{p}[0]")
        if sid > MAX_SOCKET_ID:
            _fail("invalid_socket_id", f"socket id must be 0..{MAX_SOCKET_ID} in schema v1", f"{p}[0]")
        if sid in seen:  # duplicates rejected even when one side is null
            _fail("duplicate_socket", f"duplicate socket {sid}", p)
        seen.add(sid)
        modifier = None if s[1] is None else _parse_modifier(s[1], f"{p}[1]")
        parsed.append((sid, modifier))
    parsed.sort(key=lambda pair: pair[0])  # canonical ordering
    return tuple(Socket(sid, mod) for sid, mod in parsed if mod is not None)  # empties omitted


def _parse_modifier(m: Any, path: str) -> Modifier:
    if not isinstance(m, list) or len(m) != 5 or m[0] != MODIFIER_TAG:
        _fail(
            "invalid_structure",
            f'expected ["{MODIFIER_TAG}", shape, orientation, diacritics, sockets]',
            path,
        )
    if not isinstance(m[3], list):
        _fail("invalid_structure", "diacritics must be an array", f"{path}[3]")
    return Modifier(
        _nat(m[1], f"{path}[1]"),
        _orient(m[2], f"{path}[2]"),
        tuple(_nat(d, f"{path}[3][{i}]") for i, d in enumerate(m[3])),
        _parse_sockets(m[4], f"{path}[4]"),
    )


# --------------------------------------------------------------------------
# Canonical Word -> wire (tagged arrays).
# --------------------------------------------------------------------------


def word_to_wire(word: Word) -> list:
    """Canonical wire form: plain lists, ints, tag strings -- ``json.dumps``
    of the result matches JS ``JSON.stringify`` of the same coordinate."""
    return [WORD_TAG, word.version, [_glyph_wire(g) for g in word.glyphs]]


def _glyph_wire(g: Glyph) -> list:
    return [GLYPH_TAG, g.character_class, g.base, g.orientation, [_socket_wire(s) for s in g.sockets]]


def _socket_wire(s: Socket) -> list:
    return [s.socket_id, _modifier_wire(s.modifier)]


def _modifier_wire(m: Modifier) -> list:
    return [MODIFIER_TAG, m.shape, m.orientation, list(m.diacritics), [_socket_wire(s) for s in m.sockets]]


def canonical_wire(value: Union[str, list, Word]) -> list:
    """Any accepted input form -> canonical wire form."""
    return word_to_wire(word_from_wire(value))


def as_word(value: Union[str, list, Word]) -> Word:
    """Coerce ``Word`` | wire list | JSON text to a canonical :class:`Word`."""
    return word_from_wire(value)
